### ISSUE TITLE
build failed due to a error thrown by line 126-133 of `model-building.Rmd`

### ISSUE THREAD

**Issue by dongzhuoer** [hadley/r4ds#612]
````
```{r}
grid <- diamonds2 %>% 
  data_grid(cut, .model = mod_diamond2) %>% 
  add_predictions(mod_diamond2)
grid

ggplot(grid, aes(cut, pred)) + 
  geom_point()
```
````

when I debug `modelr::data_grid()`, I find the error lies in 
```r
typical <- tidyr::crossing_(lapply(data[needed], typical))
```

refer to #608

**Comment by agwells** [comment]
As a workaround, I found that I was able to compile the book if I downgraded to tidyr version 0.6.2.

**Comment by dongzhuoer** [comment]
Hi, I found some clues:

```r
> data_grid(diamonds2, cut, .model = lm(lprice ~ lcarat + cut, data = diamonds2))
# A tibble: 5 x 2
  cut       lcarat
  <ord>      <dbl>
1 Fair      -0.515
2 Good      -0.515
3 Very Good -0.515
4 Premium   -0.515
5 Ideal     -0.515
> data_grid(diamonds2, cut, .model = lm(lprice ~ lcarat + color + cut, data = diamonds2))
Error in eval_tidy(xs[[i]], unique_output) : object 'G' not found
> data_grid(diamonds2, cut, .model = lm(lprice ~ lcarat + cut + + clarity, data = diamonds2))
Error in eval_tidy(xs[[i]], unique_output) : object 'SI1' not found
```

**Comment by dongzhuoer** [comment]
```r
> devtools::session_info()
Session info --------------------------------------------------------------------------------------------------------------------------
 setting  value                       
 version  R version 3.4.3 (2017-11-30)
 system   x86_64, linux-gnu           
 ui       RStudio (1.1.383)           
 language en_US                       
 collate  en_US.UTF-8                 
 tz       PRC                         
 date     2018-01-20                  

Packages ------------------------------------------------------------------------------------------------------------------------------
 package      * version    date       source                             
 assertthat     0.2.0      2017-04-11 CRAN (R 3.4.2)                     
 backports      1.1.2      2017-12-13 CRAN (R 3.4.3)                     
 base         * 3.4.3      2017-12-01 local                              
 bindr          0.1        2016-11-13 CRAN (R 3.4.2)                     
 bindrcpp     * 0.2        2017-06-17 CRAN (R 3.4.2)                     
 broom          0.4.3      2017-11-20 CRAN (R 3.4.2)                     
 cellranger     1.1.0      2016-07-27 CRAN (R 3.4.2)                     
 cli            1.0.0      2017-11-04 Github (r-lib/boxes@ab1c3aa)       
 colorspace     1.3-2      2016-12-14 CRAN (R 3.4.2)                     
 compiler       3.4.3      2017-12-01 local                              
 crayon         1.3.4      2017-10-31 Github (gaborcsardi/crayon@b5221ab)
 datasets     * 3.4.3      2017-12-01 local                              
 devtools       1.13.4     2017-11-09 CRAN (R 3.4.2)                     
 digest         0.6.13     2017-12-14 CRAN (R 3.4.3)                     
 dplyr        * 0.7.4.9000 2018-01-13 Github (tidyverse/dplyr@f49115b)   
 evaluate       0.10.1     2017-06-24 CRAN (R 3.4.2)                     
 forcats      * 0.2.0      2017-01-23 CRAN (R 3.4.2)                     
 foreign        0.8-69     2017-06-21 CRAN (R 3.4.2)                     
 ggplot2      * 2.2.1.9000 2018-01-13 Github (tidyverse/ggplot2@bf9255e) 
 glue           1.2.0      2017-10-29 CRAN (R 3.4.2)                     
 graphics     * 3.4.3      2017-12-01 local                              
 grDevices    * 3.4.3      2017-12-01 local                              
 grid           3.4.3      2017-12-01 local                              
 gtable         0.2.0      2016-02-26 CRAN (R 3.4.2)                     
 haven          1.1.0      2017-07-09 CRAN (R 3.4.2)                     
 hexbin       * 1.27.1     2016-12-05 cran (@1.27.1)                     
 hms            0.4.0      2017-11-23 CRAN (R 3.4.2)                     
 htmltools      0.3.6      2017-04-28 CRAN (R 3.4.2)                     
 httr           1.3.1      2017-08-20 CRAN (R 3.4.2)                     
 jsonlite       1.5        2017-06-01 CRAN (R 3.4.2)                     
 knitr          1.18       2017-12-27 CRAN (R 3.4.3)                     
 labeling       0.3        2014-08-23 CRAN (R 3.4.2)                     
 lattice        0.20-35    2017-03-25 CRAN (R 3.4.2)                     
 lazyeval       0.2.1      2017-10-29 CRAN (R 3.4.2)                     
 lubridate    * 1.7.1      2017-11-03 cran (@1.7.1)                      
 magrittr     * 1.5        2014-11-22 CRAN (R 3.4.2)                     
 MASS           7.3-47     2017-04-21 CRAN (R 3.4.2)                     
 memoise        1.1.0      2017-04-21 CRAN (R 3.4.2)                     
 methods      * 3.4.3      2017-12-01 local                              
 mnormt         1.5-5      2016-10-15 CRAN (R 3.4.2)                     
 modelr       * 0.1.1      2017-07-24 CRAN (R 3.4.2)                     
 munsell        0.4.3      2016-02-13 CRAN (R 3.4.2)                     
 nlme           3.1-131    2017-02-06 CRAN (R 3.4.2)                     
 nycflights13 * 0.2.2      2017-01-27 cran (@0.2.2)                      
 parallel       3.4.3      2017-12-01 local                              
 pillar         1.0.1      2017-11-27 CRAN (R 3.4.3)                     
 pkgconfig      2.0.1      2017-03-21 CRAN (R 3.4.2)                     
 plyr           1.8.4      2016-06-08 CRAN (R 3.4.2)                     
 psych          1.7.8      2017-09-09 CRAN (R 3.4.2)                     
 purrr        * 0.2.4      2017-10-18 CRAN (R 3.4.2)                     
 R6             2.2.2      2017-06-17 CRAN (R 3.4.2)                     
 Rcpp           0.12.14    2017-11-23 CRAN (R 3.4.2)                     
 readr        * 1.1.1.9000 2017-12-29 Github (tidyverse/readr@5415cc1)   
 readxl         1.0.0      2017-04-18 CRAN (R 3.4.2)                     
 reshape2       1.4.3      2017-12-11 CRAN (R 3.4.3)                     
 rlang          0.1.6.9003 2018-01-13 Github (tidyverse/rlang@6a4365b)   
 rmarkdown      1.8        2017-11-17 CRAN (R 3.4.2)                     
 rprojroot      1.3-2      2018-01-03 CRAN (R 3.4.3)                     
 rstudioapi     0.7        2017-09-07 CRAN (R 3.4.2)                     
 rvest          0.3.2.9000 2017-11-05 Github (hadley/rvest@9a51a5d)      
 scales         0.5.0.9000 2017-11-04 Github (hadley/scales@d767915)     
 splines      * 3.4.3      2017-12-01 local                              
 stats        * 3.4.3      2017-12-01 local                              
 stringi        1.1.6      2017-11-17 CRAN (R 3.4.2)                     
 stringr      * 1.2.0      2017-02-18 CRAN (R 3.4.2)                     
 tibble       * 1.4.1      2017-12-25 CRAN (R 3.4.3)                     
 tidyr        * 0.7.2      2017-10-16 CRAN (R 3.4.2)                     
 tidyselect     0.2.3      2017-11-06 CRAN (R 3.4.2)                     
 tidyverse    * 1.2.1      2017-11-14 CRAN (R 3.4.2)                     
 tools          3.4.3      2017-12-01 local                              
 utf8           1.1.3      2018-01-03 CRAN (R 3.4.3)                     
 utils        * 3.4.3      2017-12-01 local                              
 viridisLite    0.2.0      2017-03-24 CRAN (R 3.4.2)                     
 withr          2.1.1.9000 2018-01-13 Github (jimhester/withr@df18523)   
 xml2           1.1.1      2017-01-24 CRAN (R 3.4.2)                     
 yaml           2.1.16     2017-12-12 CRAN (R 3.4.3)
```

**Comment by batpigandme** [comment]
Adding a reprex with some comments added re. location in book, since I made it anyhow for the community discussion thread: [**Error replicating r4ds book example**](https://community.rstudio.com/t/error-replicating-r4ds-book-example/5389
).
``` r
suppressPackageStartupMessages(library(tidyverse))
library(modelr)

## R4DS: 24.2.1 Price and carat
diamonds2 <- diamonds %>% 
  filter(carat <= 2.5) %>% 
  mutate(lprice = log2(price), lcarat = log2(carat))

mod_diamond <- lm(lprice ~ lcarat, data = diamonds2)

grid <- diamonds2 %>% 
  data_grid(carat = seq_range(carat, 20)) %>% 
  mutate(lcarat = log2(carat)) %>% 
  add_predictions(mod_diamond, "lprice") %>% 
  mutate(price = 2 ^ lprice)

diamonds2 <- diamonds2 %>% 
  add_residuals(mod_diamond, "lresid")

## R4DS: 24.2.2 A more complicated model
mod_diamond2 <- lm(lprice ~ lcarat + color + cut + clarity, data = diamonds2)

grid <- diamonds2 %>% 
  data_grid(cut, .model = mod_diamond2) %>% 
  add_predictions(mod_diamond2)
#> Error in eval_tidy(xs[[i]], unique_output): object 'G' not found
```

Created on 2018-02-19 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by coatless** [comment]
Just add `eval = FALSE` to the code chunk and everything should compile.

**Comment by hadley** [comment]
Fixed in modelr 0.1.2
